src/coreclr/vm/gcheaputilities.h

```diff
@@ -5,13 +5,65 @@
 #define _GCHEAPUTILITIES_H_
 
 #include "gcinterface.h"
+#include "math.h"
 
 // The singular heap instance.
 GPTR_DECL(IGCHeap, g_pGCHeap);
 
 #ifndef DACCESS_COMPILE
 extern "C" {
 #endif // !DACCESS_COMPILE
+
+// This struct adds some state that is only visible to the EE onto the standard gc_alloc_context
+typedef struct _ee_alloc_context
+{
+    // This is the address where the next allocation sampling should occur:
+    //  - if sampling is off, this is the same as alloc_limit
+    //  - if sampling is on, this is the next sampled byte in the gc_alloc_context
+    //  - if no such byte exists (i.e. the sampling threshold is outside of the allocation context),
+    //     this is the same as alloc_limit
+    uint8_t* alloc_sampling;
+    gc_alloc_context gc_alloc_context;
+
+    // TODO: how to get a random number?
+    // - have a dedicated CLRRandom instance: seems like a waste because the related Thread already has its CLRRandom instance
+    // - share the CLRRandom instance of the related thread but the Init() method of the allocation context runs BEFORE the thread's m_random.Init() is called
```

Comment:
I wanted to avoid getting the current thread every time ComputeSamplingLimit() is called. Instead, "sharing" its CLRRandom instance by storing it in m_pRandom.

Reply:
It should be quite cheap to invoke GetThread(), we would have just accessed it recently to get the pointer to the AC so it should still be in the L1 cache. If we need to micro-optimize beyond that lets defer it to the end when we have a performance trace in hand.